File: php-backend/src/Services/ConfigExport.php
<?php

declare(strict_types=1);

namespace App\Services;

use App\Config;
use App\Database;

/**
 * 报表配置 JSON 导出，与 Python 版 services/config_export.py 对等。
 */
final class ConfigExport
{
    public const EXPORT_VERSION = '1.0';

    private const SETTINGS_KEYS = [
        'order_file',
        'order_sheet',
        'order_date_col',
        'order_date_format',
        'order_id_col',
        'sku_id_col',
        'daily_generate_at',
        'sample_rule',
        'meta',
        'excel_template_file',
    ];

    private static function serializePart(array $part): array
    {
        return [
            'sort_order' => (int) ($part['sort_order'] ?? 0),
            'label' => $part['label'] ?? null,
            'ref_field_code' => $part['ref_field_code'] ?? null,
            'source_file_keyword' => $part['source_file_keyword'] ?? null,
            'sheet_name' => $part['sheet_name'] ?? null,
            'column_header' => $part['column_header'] ?? null,
            'aliases' => Database::jsonDecode($part['aliases'] ?? null, []) ?: [],
            'sources' => Database::jsonDecode($part['sources'] ?? null, []) ?: [],
            'combine_op' => $part['combine_op'] ?? 'add',
            'aggregation' => $part['aggregation'] ?? 'sum',
            'dedup_keys' => Database::jsonDecode($part['dedup_keys'] ?? null, []) ?: [],
            'date_filter_column' => $part['date_filter_column'] ?? null,
            'date_format' => $part['date_format'] ?? null,
            'row_filters' => Database::jsonDecode($part['row_filters'] ?? null, []) ?: [],
            'exclude_sample' => (bool) ($part['exclude_sample'] ?? false),
            'exclude_review' => (bool) ($part['exclude_review'] ?? false),
            'join_to_orders' => (bool) ($part['join_to_orders'] ?? false),
            'join_keys' => Database::jsonDecode($part['join_keys'] ?? null, []) ?: [],
            'benchmark_keys' => Database::jsonDecode($part['benchmark_keys'] ?? null, []) ?: [],
            'only_sample' => (bool) ($part['only_sample'] ?? false),
        ];
    }

    private static function serializeReportLine(array $mapping): array
    {
        $lineType = $mapping['line_type'] ?? null;
        if (!$lineType) {
            if (MappingUtils::isFormulaLine($mapping)) {
                $lineType = 'formula';
            } elseif (MappingUtils::isManualLine($mapping)) {
                $lineType = 'manual';
            } else {
                $lineType = 'fetch';
            }
        }
        $parts = $mapping['parts'] ?? [];
        usort($parts, fn($a, $b) => ($a['sort_order'] ?? 0) <=> ($b['sort_order'] ?? 0));
        return [
            'line_type' => $lineType,
            'line_code' => MappingUtils::mappingLineCode($mapping),
            'label' => MappingUtils::mappingLabel($mapping),
            'report_group' => $mapping['report_group'] ?? null,
            'sort_order' => (int) ($mapping['sort_order'] ?? 0),
            'expression' => $mapping['expression'] ?? null,
            'format_type' => ($mapping['format_type'] ?? '') ?: 'usd',
            'is_highlight' => (bool) ($mapping['is_highlight'] ?? false),
            'description' => $mapping['description'] ?? null,
            'logical_field_code' => $mapping['logical_field_code'] ?? null,
            'parts' => array_map(fn($p) => self::serializePart($p), $parts),
        ];
    }

    private static function exportCatalog(int $dataSourceId): array
    {
        $filesOut = [];
        $files = Database::fetchAll(
            'SELECT * FROM catalog_files WHERE data_source_id = ? AND is_active = 1 ORDER BY id',
            [$dataSourceId]
        );
        foreach ($files as $file) {
            $sheetsOut = [];
            $sheets = Database::fetchAll(
                'SELECT * FROM catalog_sheets WHERE file_id = ? AND is_active = 1 ORDER BY sheet_name',
                [(int) $file['id']]
            );
            foreach ($sheets as $sheet) {
                $columns = Database::fetchAll(
                    'SELECT header_name, db_column, column_aliases, data_type FROM catalog_columns
                     WHERE sheet_id = ? AND is_active = 1 ORDER BY header_name',
                    [(int) $sheet['id']]
                );
                $columnsOut = array_map(fn($col) => [
                    'header_name' => $col['header_name'],
                    'db_column' => $col['db_column'],
                    'column_aliases' => Database::jsonDecode($col['column_aliases'] ?? null, []) ?: [],
                    'data_type' => ($col['data_type'] ?? '') ?: 'string',
                ], $columns);
                $sheetsOut[] = [
                    'sheet_name' => $sheet['sheet_name'],
                    'fact_table' => $sheet['fact_table'],
                    'columns' => $columnsOut,
                ];
            }
            $filesOut[] = [
                'keyword' => $file['keyword'],
                'file_label' => $file['file_label'],
                'file_name' => $file['file_name'],
                'sheets' => $sheetsOut,
            ];
        }
        return ['files' => $filesOut];
    }

    private static function exportSettings(array $cfg, bool $includeReviewOrders): array
    {
        $out = [];
        foreach (self::SETTINGS_KEYS as $key) {
            if (array_key_exists($key, $cfg)) {
                $out[$key] = $cfg[$key];
            }
        }
        if ($includeReviewOrders) {
            if (!empty($cfg['review_orders'])) {
                $out['review_orders'] = $cfg['review_orders'];
            } elseif (!empty($cfg['review_order_ids'])) {
                $out['review_order_ids'] = $cfg['review_order_ids'];
            }
        }
        return $out;
    }

    public static function buildConfigExport(array $ds, bool $includeReviewOrders = true): array
    {
        $cfg = DsSettings::getDsConfig($ds);
        $store = Database::fetchOne('SELECT * FROM stores WHERE data_source_id = ?', [(int) $ds['id']]);
        $mappings = MappingRepo::forDataSource((int) $ds['id'], true);
        return [
            'export_version' => self::EXPORT_VERSION,
            'exported_at' => (new \DateTimeImmutable('now', new \DateTimeZone('Asia/Shanghai')))->format('Y-m-d\TH:i:s'),
            'store' => [
                'name' => $store['name'] ?? null,
                'platform' => $store['platform'] ?? null,
            ],
            'data_source' => [
                'name' => $ds['name'],
                'platform' => $ds['platform'],
                'description' => $ds['description'] ?? null,
            ],
            'settings' => self::exportSettings($cfg, $includeReviewOrders),
            'catalog' => self::exportCatalog((int) $ds['id']),
            'report_lines' => array_map(fn($m) => self::serializeReportLine($m), $mappings),
        ];
    }

    public static function exportFilename(array $ds): string
    {
        $stamp = (new \DateTimeImmutable('now', new \DateTimeZone('Asia/Shanghai')))->format('Ymd');
        return "autoreport-config_ds{$ds['id']}_{$stamp}.json";
    }
}

File: app/services/config_export.py
"""报表配置 JSON 导出。"""

from __future__ import annotations

from typing import Any

from sqlalchemy.orm import Session, joinedload

from app.models import CatalogColumn, CatalogFile, CatalogSheet, DataSource, FieldMapping
from app.services.ds_settings import get_ds_config
from app.services.mapping_utils import is_formula_line, is_manual_line, mapping_label, mapping_line_code
from app.services.timezone import now_cst

EXPORT_VERSION = "1.0"

_SETTINGS_KEYS = (
    "order_file",
    "order_sheet",
    "order_date_col",
    "order_date_format",
    "order_id_col",
    "sku_id_col",
    "daily_generate_at",
    "sample_rule",
    "meta",
    "excel_template_file",
    "review_logistics_mode",
    "review_logistics_per_order",
    "review_logistics_exclude_same_day_refund",
)


def _serialize_part(part) -> dict[str, Any]:
    return {
        "sort_order": part.sort_order,
        "label": part.label,
        "ref_field_code": part.ref_field_code,
        "source_file_keyword": part.source_file_keyword,
        "sheet_name": part.sheet_name,
        "column_header": part.column_header,
        "aliases": list(part.aliases or []),
        "sources": list(part.sources or []),
        "combine_op": part.combine_op,
        "aggregation": part.aggregation,
        "dedup_keys": list(part.dedup_keys or []),
        "date_filter_column": part.date_filter_column,
        "date_format": part.date_format,
        "row_filters": list(part.row_filters or []),
        "exclude_sample": bool(part.exclude_sample),
        "exclude_review": bool(part.exclude_review),
        "exclude_same_day_refund": bool(getattr(part, "exclude_same_day_refund", False)),
        "join_to_orders": bool(part.join_to_orders),
        "join_keys": list(part.join_keys or []),
        "benchmark_keys": list(getattr(part, "benchmark_keys", None) or []),
        "only_sample": bool(getattr(part, "only_sample", False)),
    }


def _serialize_report_line(mapping: FieldMapping) -> dict[str, Any]:
    line_type = mapping.line_type or (
        "formula" if is_formula_line(mapping) else ("manual" if is_manual_line(mapping) else "fetch")
    )
    return {
        "line_type": line_type,
        "line_code": mapping_line_code(mapping),
        "label": mapping_label(mapping),
        "report_group": mapping.report_group,
        "sort_order": mapping.sort_order or 0,
        "expression": mapping.expression,
        "format_type": mapping.format_type or "usd",
        "is_highlight": bool(mapping.is_highlight),
        "description": mapping.description,
        "logical_field_code": mapping.logical_field.code if mapping.logical_field else None,
        "parts": [_serialize_part(p) for p in sorted(mapping.parts, key=lambda x: x.sort_order)],
    }


def _export_catalog(db: Session, data_source_id: int) -> dict[str, Any]:
    files_out: list[dict[str, Any]] = []
    rows = (
        db.query(CatalogFile)
        .options(
            joinedload(CatalogFile.sheets).joinedload(CatalogSheet.columns),
        )
        .filter(CatalogFile.data_source_id == data_source_id, CatalogFile.is_active.is_(True))
        .order_by(CatalogFile.id)
        .all()
    )
    for file in rows:
        sheets_out: list[dict[str, Any]] = []
        for sheet in sorted(file.sheets, key=lambda s: s.sheet_name):
            if not sheet.is_active:
                continue
            columns = [
                {
                    "header_name": col.header_name,
                    "db_column": col.db_column,
                    "column_aliases": list(col.column_aliases or []),
                    "data_type": col.data_type or "string",
                }
                for col in sorted(sheet.columns, key=lambda c: c.header_name)
                if col.is_active
            ]
            sheets_out.append(
                {
                    "sheet_name": sheet.sheet_name,
                    "fact_table": sheet.fact_table,
                    "columns": columns,
                }
            )
        files_out.append(
            {
                "keyword": file.keyword,
                "file_label": file.file_label,
                "file_name": file.file_name,
                "sheets": sheets_out,
            }
        )
    return {"files": files_out}


def _export_settings(cfg: dict, *, include_review_orders: bool) -> dict[str, Any]:
    out = {key: cfg.get(key) for key in _SETTINGS_KEYS if key in cfg}
    if include_review_orders:
        if cfg.get("review_orders"):
            out["review_orders"] = list(cfg["review_orders"])
        elif cfg.get("review_order_ids"):
            out["review_order_ids"] = list(cfg["review_order_ids"])
    return out


def build_config_export(
    db: Session,
    ds: DataSource,
    *,
    include_review_orders: bool = True,
) -> dict[str, Any]:
    cfg = get_ds_config(ds)
    store = ds.store
    mappings = (
        db.query(FieldMapping)
        .options(joinedload(FieldMapping.logical_field), joinedload(FieldMapping.parts))
        .filter(FieldMapping.data_source_id == ds.id)
        .order_by(FieldMapping.sort_order, FieldMapping.id)
        .all()
    )
    return {
        "export_version": EXPORT_VERSION,
        "exported_at": now_cst().isoformat(timespec="seconds"),
        "store": {
            "name": store.name if store else None,
            "platform": store.platform if store else None,
        },
        "data_source": {
            "name": ds.name,
            "platform": ds.platform,
            "description": ds.description,
        },
        "settings": _export_settings(cfg, include_review_orders=include_review_orders),
        "catalog": _export_catalog(db, ds.id),
        "report_lines": [_serialize_report_line(m) for m in mappings],
    }


def export_filename(ds: DataSource) -> str:
    """HTTP Content-Disposition 仅支持 ASCII，文件名用 ds id + 日期。"""
    stamp = now_cst().strftime("%Y%m%d")
    return f"autoreport-config_ds{ds.id}_{stamp}.json"
